pkg/oam/util/helper.go

```diff
@@ -204,33 +205,103 @@ func FetchWorkloadDefinition(ctx context.Context, r client.Reader, dm discoverym
 	return workloadDefinition, nil
 }
 
-// GenNamespacedDefinitionName generate definition name with customized namespace
-func GenNamespacedDefinitionName(dn string) types.NamespacedName {
-	if dns := os.Getenv(DefinitionNamespaceEnv); dns != "" {
-		return types.NamespacedName{Name: dn, Namespace: dns}
+// GenNamespaceWithCtx gen sysNs and appNs from ctx
+func GenNamespaceWithCtx(ctx context.Context) (string, string) {
+	var sysNs, appNs string
+	if sys := ctx.Value(sysDefinitionNamespace); sys == nil {
+		sysNs = "vela-system"
+	} else {
+		sysNs = sys.(string)
+	}
+	if app := ctx.Value(appDefinitionNamespace); app == nil {
+		appNs = "vela-system"
+	} else {
+		appNs = app.(string)
 	}
-	return types.NamespacedName{Name: dn}
+	return sysNs, appNs
+}
+
+// SetNnamespaceInCtx set system-level namespace and app namespace in context
+func SetNnamespaceInCtx(ctx context.Context, sysNs string, appNs string) context.Context {
+	ctx = context.WithValue(ctx, sysDefinitionNamespace, sysNs)
+	if appNs == "" {
+		appNs = "default"
+	}
+	ctx = context.WithValue(ctx, appDefinitionNamespace, appNs)
+	return ctx
 }
 
 // GetWorkloadDefinition  Get WorkloadDefinition
 func GetWorkloadDefinition(ctx context.Context, cli client.Reader,
 	workitemName string) (*v1alpha2.WorkloadDefinition, error) {
 	wd := new(v1alpha2.WorkloadDefinition)
-	if err := cli.Get(ctx, GenNamespacedDefinitionName(workitemName), wd); err != nil {
-		return nil, err
+	if dns := os.Getenv(DefinitionNamespaceEnv); dns != "" {
+		if err := cli.Get(ctx, types.NamespacedName{Name: workitemName, Namespace: dns}, wd); err == nil {
+			return wd, nil
+		} else if !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+	}
+	sysNs, userNs := GenNamespaceWithCtx(ctx)
+	if err := cli.Get(ctx, types.NamespacedName{Name: workitemName, Namespace: userNs}, wd); err != nil {
+		if apierrors.IsNotFound(err) {
+			if err = cli.Get(ctx, types.NamespacedName{Name: workitemName, Namespace: sysNs}, wd); err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
```

Comment:
I see the code is a little duplicate 3 times, can we try to reuse or merge to one function?